<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>smeagol.enrich API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>smeagol.enrich</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># General imports
import numpy as np
import pandas as pd
import itertools

# Stats imports
import scipy.stats as stats
import statsmodels.stats.multitest as multitest

# Smeagol imports
from .utils import shuffle_records, read_bg_seqs
from .encode import SeqGroups
from .scan import scan_sequences, get_tiling_windows_over_genome, count_in_sliding_windows, count_in_window


def enrich_over_shuffled(real_counts, shuf_stats, background=&#39;binomial&#39;, records=None):
    &#34;&#34;&#34;Function to calculate enrichment of binding sites in real vs. shuffled genomes
    
    Args:
        real_counts (pd.DataFrame): counts of binding sites in real genome
        shuf_stats (pd.DataFrame): statistics for binding sites across multiple shuffled genomes
        background (str): &#39;binomial&#39; or &#39;normal&#39;
        records (SeqGroups): SeqGroups object. Only needed if background=&#39;binomial&#39;
        
    Returns:
        enr_full (pd.DataFrame): dataframe containing FDR-corrected p-values for enrichment of each PWM.  
        
    &#34;&#34;&#34;
    # Compare counts on real and shuffled genomes for each PWM
    if &#39;name&#39; in shuf_stats.columns:
        enr = real_counts.merge(shuf_stats, on=[&#39;Matrix_id&#39;, &#39;width&#39;, &#39;sense&#39;, &#39;name&#39;], how=&#39;outer&#39;)
    else:
        enr = real_counts.merge(shuf_stats, on=[&#39;Matrix_id&#39;, &#39;width&#39;, &#39;sense&#39;], how=&#39;outer&#39;)

    # If 0 sites are present in real genome, include the entry
    enr[&#39;num&#39;] = enr[&#39;num&#39;].fillna(0)

    # If 0 sites are present in shuffled genomes, set a minimum of 1 site
    num_shuf = enr[&#39;len&#39;][0]
    enr.loc[enr.avg==0, &#39;avg&#39;] = 1/num_shuf

    # Calculate normal z-score
    if (background == &#39;normal&#39;) or (background == &#39;both&#39;):
        enr.loc[enr.avg==0, &#39;sd&#39;] = np.std([1] + [0]*(num_shuf - 1))
        enr[&#39;z&#39;] = (enr.num - enr.avg)/enr.sd
        enr[&#39;z&#39;] = enr.z.replace([-np.inf], -10)

    # Calculate p-value
    if background == &#39;normal&#39;:
        enr[&#39;pnorm&#39;] = stats.norm.sf(abs(enr.z))*2
    else:
        enr[&#39;adj_len&#39;] = sum([len(record) - enr.width + 1 for record in records])
        enr[&#39;p&#39;] = enr.apply(lambda x:stats.binom_test(x[&#39;num&#39;], x[&#39;adj_len&#39;], x[&#39;avg&#39;]/x[&#39;adj_len&#39;], alternative=&#39;two-sided&#39;), axis=1)
    enr_full = pd.DataFrame()

    # FDR correction per sense
    for sense in pd.unique(enr.sense):
        enr_x = enr[enr.sense == sense].copy()
        if background == &#39;normal&#39;:
            enr_x[&#39;fdr_norm&#39;] = multitest.fdrcorrection(enr_x.pnorm)[1]  
        else:
            enr_x[&#39;fdr&#39;] = multitest.fdrcorrection(enr_x.p)[1]
        enr_full = pd.concat([enr_full, enr_x])

    # Sort and index final results
    if background == &#39;normal&#39;:
        enr_full.sort_values([&#39;sense&#39;, &#39;fdr_binom&#39;], inplace=True)
    else:
        enr_full.sort_values([&#39;sense&#39;, &#39;fdr&#39;], inplace=True)
    enr_full.reset_index(inplace=True, drop=True)

    return enr_full


def enrich_in_genome(records, model, simN, simK, rcomp, threshold, sense=&#39;+&#39;, background=&#39;binomial&#39;, verbose=False, combine_seqs=True, seq_batch=0, shuf_seqs=None):
    &#34;&#34;&#34;Function to shuffle sequence(s) and calculate enrichment of PWMs in sequence(s) relative to the shuffled background.
        
    Args:
        records (list): list of seqrecord objects
        model (PWMModel): parameterized convolutional model
        simN (int): number of shuffles
        simK (int): k-mer frequency to conserve while shuffling
        rcomp (str): &#39;only&#39; to encode the sequence reverse complements, &#39;both&#39; to encode the reverse
                     complements as well as original sequences, or &#39;none&#39;.
        sense (str): &#39;+&#39; or &#39;-&#39;        
        background (str): &#39;binomial&#39; or &#39;normal&#39;
        combine_seqs (bool): combine outputs for all sequence groups into single dataframe
        seq_batch (int): number of shuffled sequences to scan at a time. If 0, scan all.
        shuf_seqs (str/list): path to FASTA file or list of seqrecord objects containing shuffled sequences
        
    Returns:
        results (dict): dictionary containing results.  
        
    &#34;&#34;&#34;
    # Find sites on real genome
    real_preds = scan_sequences(records, model, threshold, sense, rcomp, outputs=[&#39;sites&#39;, &#39;counts&#39;], score=False, combine_seqs=combine_seqs)

    # Shuffle genome
    if shuf_seqs is not None:
        if type(shuf_seqs) == str:
            shuf = read_bg_seqs(shuf_seqs, records, simN)
        elif type(shuf_seqs) == list:
            shuf = shuf_seqs    
    else:
        shuf = shuffle_records(records, simN, simK)
    
    # Count sites on shuffled genomes
    shuf_preds = scan_sequences(shuf, model, threshold, sense, rcomp, outputs=[&#39;counts&#39;, &#39;stats&#39;], group_by=&#39;name&#39;, combine_seqs=combine_seqs, sep_ids=True, seq_batch=seq_batch)
        
    # Calculate binding site enrichment
    enr = enrich_over_shuffled(real_preds[&#39;counts&#39;], shuf_preds[&#39;stats&#39;], background=background, records=records)

    # Combine results
    results = {&#39;enrichment&#39;: enr, 
               &#39;real_sites&#39;: real_preds[&#39;sites&#39;], 
               &#39;real_counts&#39;: real_preds[&#39;counts&#39;], 
               &#39;shuf_counts&#39;: shuf_preds[&#39;counts&#39;],
               &#39;shuf_stats&#39;: shuf_preds[&#39;stats&#39;]}
    if verbose:
        
        results[&#39;shuf_seqs&#39;] = shuf
    return results


def examine_thresholds(records, model, simN, simK, rcomp, min_threshold, sense=&#39;+&#39;, verbose=False, combine_seqs=True):
    &#34;&#34;&#34;Function to compare the number of binding sites at various thresholds.
            
    Args:
        records (list): list of seqrecord objects
        model (PWMModel): parameterized convolutional model
        simN (int): number of shuffles
        simK (int): k-mer frequency to conserve while shuffling
        rcomp (str): &#39;only&#39; to encode the sequence reverse complements, &#39;both&#39; to encode the reverse
                     complements as well as original sequences, or &#39;none&#39;.
        sense (str): &#39;+&#39; or &#39;-&#39;        
        min_threshold (float): minimum threshold for a binding site (0 to 1)
        verbose (bool): output all information
        combine_seqs (bool): combine outputs for multiple sequences into single dataframe
        
    Returns:
        results (dict): dictionary containing results. 
    &#34;&#34;&#34;
    thresholds = np.arange(min_threshold, 1.0, 0.1)
    
    # scan real sequence
    real_binned = scan_sequences(records, model, thresholds, sense, rcomp, outputs=[&#39;binned_counts&#39;], combine_seqs=combine_seqs)[&#39;binned_counts&#39;]
    
    # shuffle
    shuf = shuffle_records(records, simN, simK)
    
    # scan shuffled sequence
    shuf_binned = scan_sequences(shuf, model, thresholds, sense, rcomp, outputs=[&#39;binned_counts&#39;], combine_seqs=combine_seqs, group_by=&#39;name&#39;, sep_ids=True)[&#39;binned_counts&#39;]
    
    results = {&#39;real_binned&#39;:real_binned, &#39;shuf_binned&#39;: shuf_binned}
    if verbose:
        results[&#39;shuf_seqs&#39;] = shuf
    return results


def enrich_in_window(window, sites, genome, matrix_id):
    &#34;&#34;&#34;Function to calculate enrichment of PWMs in subsequence relative to total sequence.
    
    Args:
        window (pd.DataFrame): dataframe with columns id, start, end
        sites (pd.DataFrame): dataframe containing locations of binding sites
        genome (list): list of SeqRecord objects
        matrix_id (str): selected PWM
        
    Returns:
        result (pd.DataFrame): Dataframe containing result of enrichment analysis in selected window.
    
    &#34;&#34;&#34;
    result = window.copy()
    # Get sites for required matrix
    matrix_sites = sites[sites.Matrix_id == matrix_id].reset_index(drop=True)
    # Get matrix width
    width = matrix_sites.width[0]
    # Correct window length and genome length
    effective_window_len = window.end - window.start - width + 1
    effective_genome_len = sum([len(x) for x in genome]) - width + 1
    # Calculate number of successes and failures in selected region
    result[&#39;len&#39;] = effective_window_len
    result[&#39;count&#39;] = count_in_window(window, sites, matrix_id)
    count_neg = result[&#39;len&#39;] - result[&#39;count&#39;]
    # Calculate number of successes and failures in whole genome
    result[&#39;tot_count&#39;] = len(matrix_sites)
    tot_neg = effective_genome_len - result.tot_count
    # Expected value
    result[&#39;expected&#39;] = (result.tot_count * result.len) / effective_genome_len
    # Fisher&#39;s exact test
    odds, p = stats.fisher_exact([[result[&#39;count&#39;], count_neg], [result.tot_count, tot_neg]], alternative=&#39;two-sided&#39;)
    result[&#39;odds&#39;] = odds
    result[&#39;p&#39;] = p
    return result


def enrich_in_sliding_windows(sites, genome, matrix_id, width, shift=None):
    &#34;&#34;&#34;Function to test enrichment of binding sites in sliding windows across a genome.
     
    Args:
        sites (pd.DataFrame): dataframe containing locations of binding sites
        genome (list): list of SeqRecord objects
        matrix_id (str): selected PWM
        width (int): width of tiling windows
        shift (int): shift between successive windows
    
    Returns:
        windows (pd.DataFrame): results of enrichment analysis in windows covering all sequences in genome
        
    &#34;&#34;&#34;
    # Get sliding windows across genome
    windows = get_tiling_windows_over_genome(genome, width, shift)
    # Perform enrichment in each window
    windows = windows.apply(enrich_in_window, axis=1, args=(sites, genome, matrix_id)).reset_index(drop=True)
    # Overall FDR correction
    windows[&#39;padj&#39;] = multitest.fdrcorrection(windows.p)[1]
    return windows</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="smeagol.enrich.enrich_in_genome"><code class="name flex">
<span>def <span class="ident">enrich_in_genome</span></span>(<span>records, model, simN, simK, rcomp, threshold, sense='+', background='binomial', verbose=False, combine_seqs=True, seq_batch=0, shuf_seqs=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to shuffle sequence(s) and calculate enrichment of PWMs in sequence(s) relative to the shuffled background.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>records</code></strong> :&ensp;<code>list</code></dt>
<dd>list of seqrecord objects</dd>
<dt><strong><code>model</code></strong> :&ensp;<code>PWMModel</code></dt>
<dd>parameterized convolutional model</dd>
<dt><strong><code>simN</code></strong> :&ensp;<code>int</code></dt>
<dd>number of shuffles</dd>
<dt><strong><code>simK</code></strong> :&ensp;<code>int</code></dt>
<dd>k-mer frequency to conserve while shuffling</dd>
<dt><strong><code>rcomp</code></strong> :&ensp;<code>str</code></dt>
<dd>'only' to encode the sequence reverse complements, 'both' to encode the reverse
complements as well as original sequences, or 'none'.</dd>
<dt><strong><code>sense</code></strong> :&ensp;<code>str</code></dt>
<dd>'+' or '-'
</dd>
<dt><strong><code>background</code></strong> :&ensp;<code>str</code></dt>
<dd>'binomial' or 'normal'</dd>
<dt><strong><code>combine_seqs</code></strong> :&ensp;<code>bool</code></dt>
<dd>combine outputs for all sequence groups into single dataframe</dd>
<dt><strong><code>seq_batch</code></strong> :&ensp;<code>int</code></dt>
<dd>number of shuffled sequences to scan at a time. If 0, scan all.</dd>
</dl>
<p>shuf_seqs (str/list): path to FASTA file or list of seqrecord objects containing shuffled sequences</p>
<h2 id="returns">Returns</h2>
<p>results (dict): dictionary containing results.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def enrich_in_genome(records, model, simN, simK, rcomp, threshold, sense=&#39;+&#39;, background=&#39;binomial&#39;, verbose=False, combine_seqs=True, seq_batch=0, shuf_seqs=None):
    &#34;&#34;&#34;Function to shuffle sequence(s) and calculate enrichment of PWMs in sequence(s) relative to the shuffled background.
        
    Args:
        records (list): list of seqrecord objects
        model (PWMModel): parameterized convolutional model
        simN (int): number of shuffles
        simK (int): k-mer frequency to conserve while shuffling
        rcomp (str): &#39;only&#39; to encode the sequence reverse complements, &#39;both&#39; to encode the reverse
                     complements as well as original sequences, or &#39;none&#39;.
        sense (str): &#39;+&#39; or &#39;-&#39;        
        background (str): &#39;binomial&#39; or &#39;normal&#39;
        combine_seqs (bool): combine outputs for all sequence groups into single dataframe
        seq_batch (int): number of shuffled sequences to scan at a time. If 0, scan all.
        shuf_seqs (str/list): path to FASTA file or list of seqrecord objects containing shuffled sequences
        
    Returns:
        results (dict): dictionary containing results.  
        
    &#34;&#34;&#34;
    # Find sites on real genome
    real_preds = scan_sequences(records, model, threshold, sense, rcomp, outputs=[&#39;sites&#39;, &#39;counts&#39;], score=False, combine_seqs=combine_seqs)

    # Shuffle genome
    if shuf_seqs is not None:
        if type(shuf_seqs) == str:
            shuf = read_bg_seqs(shuf_seqs, records, simN)
        elif type(shuf_seqs) == list:
            shuf = shuf_seqs    
    else:
        shuf = shuffle_records(records, simN, simK)
    
    # Count sites on shuffled genomes
    shuf_preds = scan_sequences(shuf, model, threshold, sense, rcomp, outputs=[&#39;counts&#39;, &#39;stats&#39;], group_by=&#39;name&#39;, combine_seqs=combine_seqs, sep_ids=True, seq_batch=seq_batch)
        
    # Calculate binding site enrichment
    enr = enrich_over_shuffled(real_preds[&#39;counts&#39;], shuf_preds[&#39;stats&#39;], background=background, records=records)

    # Combine results
    results = {&#39;enrichment&#39;: enr, 
               &#39;real_sites&#39;: real_preds[&#39;sites&#39;], 
               &#39;real_counts&#39;: real_preds[&#39;counts&#39;], 
               &#39;shuf_counts&#39;: shuf_preds[&#39;counts&#39;],
               &#39;shuf_stats&#39;: shuf_preds[&#39;stats&#39;]}
    if verbose:
        
        results[&#39;shuf_seqs&#39;] = shuf
    return results</code></pre>
</details>
</dd>
<dt id="smeagol.enrich.enrich_in_sliding_windows"><code class="name flex">
<span>def <span class="ident">enrich_in_sliding_windows</span></span>(<span>sites, genome, matrix_id, width, shift=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to test enrichment of binding sites in sliding windows across a genome.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>sites</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>dataframe containing locations of binding sites</dd>
<dt><strong><code>genome</code></strong> :&ensp;<code>list</code></dt>
<dd>list of SeqRecord objects</dd>
<dt><strong><code>matrix_id</code></strong> :&ensp;<code>str</code></dt>
<dd>selected PWM</dd>
<dt><strong><code>width</code></strong> :&ensp;<code>int</code></dt>
<dd>width of tiling windows</dd>
<dt><strong><code>shift</code></strong> :&ensp;<code>int</code></dt>
<dd>shift between successive windows</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>windows (pd.DataFrame): results of enrichment analysis in windows covering all sequences in genome</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def enrich_in_sliding_windows(sites, genome, matrix_id, width, shift=None):
    &#34;&#34;&#34;Function to test enrichment of binding sites in sliding windows across a genome.
     
    Args:
        sites (pd.DataFrame): dataframe containing locations of binding sites
        genome (list): list of SeqRecord objects
        matrix_id (str): selected PWM
        width (int): width of tiling windows
        shift (int): shift between successive windows
    
    Returns:
        windows (pd.DataFrame): results of enrichment analysis in windows covering all sequences in genome
        
    &#34;&#34;&#34;
    # Get sliding windows across genome
    windows = get_tiling_windows_over_genome(genome, width, shift)
    # Perform enrichment in each window
    windows = windows.apply(enrich_in_window, axis=1, args=(sites, genome, matrix_id)).reset_index(drop=True)
    # Overall FDR correction
    windows[&#39;padj&#39;] = multitest.fdrcorrection(windows.p)[1]
    return windows</code></pre>
</details>
</dd>
<dt id="smeagol.enrich.enrich_in_window"><code class="name flex">
<span>def <span class="ident">enrich_in_window</span></span>(<span>window, sites, genome, matrix_id)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to calculate enrichment of PWMs in subsequence relative to total sequence.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>window</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>dataframe with columns id, start, end</dd>
<dt><strong><code>sites</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>dataframe containing locations of binding sites</dd>
<dt><strong><code>genome</code></strong> :&ensp;<code>list</code></dt>
<dd>list of SeqRecord objects</dd>
<dt><strong><code>matrix_id</code></strong> :&ensp;<code>str</code></dt>
<dd>selected PWM</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>result (pd.DataFrame): Dataframe containing result of enrichment analysis in selected window.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def enrich_in_window(window, sites, genome, matrix_id):
    &#34;&#34;&#34;Function to calculate enrichment of PWMs in subsequence relative to total sequence.
    
    Args:
        window (pd.DataFrame): dataframe with columns id, start, end
        sites (pd.DataFrame): dataframe containing locations of binding sites
        genome (list): list of SeqRecord objects
        matrix_id (str): selected PWM
        
    Returns:
        result (pd.DataFrame): Dataframe containing result of enrichment analysis in selected window.
    
    &#34;&#34;&#34;
    result = window.copy()
    # Get sites for required matrix
    matrix_sites = sites[sites.Matrix_id == matrix_id].reset_index(drop=True)
    # Get matrix width
    width = matrix_sites.width[0]
    # Correct window length and genome length
    effective_window_len = window.end - window.start - width + 1
    effective_genome_len = sum([len(x) for x in genome]) - width + 1
    # Calculate number of successes and failures in selected region
    result[&#39;len&#39;] = effective_window_len
    result[&#39;count&#39;] = count_in_window(window, sites, matrix_id)
    count_neg = result[&#39;len&#39;] - result[&#39;count&#39;]
    # Calculate number of successes and failures in whole genome
    result[&#39;tot_count&#39;] = len(matrix_sites)
    tot_neg = effective_genome_len - result.tot_count
    # Expected value
    result[&#39;expected&#39;] = (result.tot_count * result.len) / effective_genome_len
    # Fisher&#39;s exact test
    odds, p = stats.fisher_exact([[result[&#39;count&#39;], count_neg], [result.tot_count, tot_neg]], alternative=&#39;two-sided&#39;)
    result[&#39;odds&#39;] = odds
    result[&#39;p&#39;] = p
    return result</code></pre>
</details>
</dd>
<dt id="smeagol.enrich.enrich_over_shuffled"><code class="name flex">
<span>def <span class="ident">enrich_over_shuffled</span></span>(<span>real_counts, shuf_stats, background='binomial', records=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to calculate enrichment of binding sites in real vs. shuffled genomes</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>real_counts</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>counts of binding sites in real genome</dd>
<dt><strong><code>shuf_stats</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>statistics for binding sites across multiple shuffled genomes</dd>
<dt><strong><code>background</code></strong> :&ensp;<code>str</code></dt>
<dd>'binomial' or 'normal'</dd>
<dt><strong><code>records</code></strong> :&ensp;<code>SeqGroups</code></dt>
<dd>SeqGroups object. Only needed if background='binomial'</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>enr_full (pd.DataFrame): dataframe containing FDR-corrected p-values for enrichment of each PWM.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def enrich_over_shuffled(real_counts, shuf_stats, background=&#39;binomial&#39;, records=None):
    &#34;&#34;&#34;Function to calculate enrichment of binding sites in real vs. shuffled genomes
    
    Args:
        real_counts (pd.DataFrame): counts of binding sites in real genome
        shuf_stats (pd.DataFrame): statistics for binding sites across multiple shuffled genomes
        background (str): &#39;binomial&#39; or &#39;normal&#39;
        records (SeqGroups): SeqGroups object. Only needed if background=&#39;binomial&#39;
        
    Returns:
        enr_full (pd.DataFrame): dataframe containing FDR-corrected p-values for enrichment of each PWM.  
        
    &#34;&#34;&#34;
    # Compare counts on real and shuffled genomes for each PWM
    if &#39;name&#39; in shuf_stats.columns:
        enr = real_counts.merge(shuf_stats, on=[&#39;Matrix_id&#39;, &#39;width&#39;, &#39;sense&#39;, &#39;name&#39;], how=&#39;outer&#39;)
    else:
        enr = real_counts.merge(shuf_stats, on=[&#39;Matrix_id&#39;, &#39;width&#39;, &#39;sense&#39;], how=&#39;outer&#39;)

    # If 0 sites are present in real genome, include the entry
    enr[&#39;num&#39;] = enr[&#39;num&#39;].fillna(0)

    # If 0 sites are present in shuffled genomes, set a minimum of 1 site
    num_shuf = enr[&#39;len&#39;][0]
    enr.loc[enr.avg==0, &#39;avg&#39;] = 1/num_shuf

    # Calculate normal z-score
    if (background == &#39;normal&#39;) or (background == &#39;both&#39;):
        enr.loc[enr.avg==0, &#39;sd&#39;] = np.std([1] + [0]*(num_shuf - 1))
        enr[&#39;z&#39;] = (enr.num - enr.avg)/enr.sd
        enr[&#39;z&#39;] = enr.z.replace([-np.inf], -10)

    # Calculate p-value
    if background == &#39;normal&#39;:
        enr[&#39;pnorm&#39;] = stats.norm.sf(abs(enr.z))*2
    else:
        enr[&#39;adj_len&#39;] = sum([len(record) - enr.width + 1 for record in records])
        enr[&#39;p&#39;] = enr.apply(lambda x:stats.binom_test(x[&#39;num&#39;], x[&#39;adj_len&#39;], x[&#39;avg&#39;]/x[&#39;adj_len&#39;], alternative=&#39;two-sided&#39;), axis=1)
    enr_full = pd.DataFrame()

    # FDR correction per sense
    for sense in pd.unique(enr.sense):
        enr_x = enr[enr.sense == sense].copy()
        if background == &#39;normal&#39;:
            enr_x[&#39;fdr_norm&#39;] = multitest.fdrcorrection(enr_x.pnorm)[1]  
        else:
            enr_x[&#39;fdr&#39;] = multitest.fdrcorrection(enr_x.p)[1]
        enr_full = pd.concat([enr_full, enr_x])

    # Sort and index final results
    if background == &#39;normal&#39;:
        enr_full.sort_values([&#39;sense&#39;, &#39;fdr_binom&#39;], inplace=True)
    else:
        enr_full.sort_values([&#39;sense&#39;, &#39;fdr&#39;], inplace=True)
    enr_full.reset_index(inplace=True, drop=True)

    return enr_full</code></pre>
</details>
</dd>
<dt id="smeagol.enrich.examine_thresholds"><code class="name flex">
<span>def <span class="ident">examine_thresholds</span></span>(<span>records, model, simN, simK, rcomp, min_threshold, sense='+', verbose=False, combine_seqs=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to compare the number of binding sites at various thresholds.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>records</code></strong> :&ensp;<code>list</code></dt>
<dd>list of seqrecord objects</dd>
<dt><strong><code>model</code></strong> :&ensp;<code>PWMModel</code></dt>
<dd>parameterized convolutional model</dd>
<dt><strong><code>simN</code></strong> :&ensp;<code>int</code></dt>
<dd>number of shuffles</dd>
<dt><strong><code>simK</code></strong> :&ensp;<code>int</code></dt>
<dd>k-mer frequency to conserve while shuffling</dd>
<dt><strong><code>rcomp</code></strong> :&ensp;<code>str</code></dt>
<dd>'only' to encode the sequence reverse complements, 'both' to encode the reverse
complements as well as original sequences, or 'none'.</dd>
<dt><strong><code>sense</code></strong> :&ensp;<code>str</code></dt>
<dd>'+' or '-'
</dd>
<dt><strong><code>min_threshold</code></strong> :&ensp;<code>float</code></dt>
<dd>minimum threshold for a binding site (0 to 1)</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code></dt>
<dd>output all information</dd>
<dt><strong><code>combine_seqs</code></strong> :&ensp;<code>bool</code></dt>
<dd>combine outputs for multiple sequences into single dataframe</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>results (dict): dictionary containing results.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def examine_thresholds(records, model, simN, simK, rcomp, min_threshold, sense=&#39;+&#39;, verbose=False, combine_seqs=True):
    &#34;&#34;&#34;Function to compare the number of binding sites at various thresholds.
            
    Args:
        records (list): list of seqrecord objects
        model (PWMModel): parameterized convolutional model
        simN (int): number of shuffles
        simK (int): k-mer frequency to conserve while shuffling
        rcomp (str): &#39;only&#39; to encode the sequence reverse complements, &#39;both&#39; to encode the reverse
                     complements as well as original sequences, or &#39;none&#39;.
        sense (str): &#39;+&#39; or &#39;-&#39;        
        min_threshold (float): minimum threshold for a binding site (0 to 1)
        verbose (bool): output all information
        combine_seqs (bool): combine outputs for multiple sequences into single dataframe
        
    Returns:
        results (dict): dictionary containing results. 
    &#34;&#34;&#34;
    thresholds = np.arange(min_threshold, 1.0, 0.1)
    
    # scan real sequence
    real_binned = scan_sequences(records, model, thresholds, sense, rcomp, outputs=[&#39;binned_counts&#39;], combine_seqs=combine_seqs)[&#39;binned_counts&#39;]
    
    # shuffle
    shuf = shuffle_records(records, simN, simK)
    
    # scan shuffled sequence
    shuf_binned = scan_sequences(shuf, model, thresholds, sense, rcomp, outputs=[&#39;binned_counts&#39;], combine_seqs=combine_seqs, group_by=&#39;name&#39;, sep_ids=True)[&#39;binned_counts&#39;]
    
    results = {&#39;real_binned&#39;:real_binned, &#39;shuf_binned&#39;: shuf_binned}
    if verbose:
        results[&#39;shuf_seqs&#39;] = shuf
    return results</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="smeagol" href="index.html">smeagol</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="smeagol.enrich.enrich_in_genome" href="#smeagol.enrich.enrich_in_genome">enrich_in_genome</a></code></li>
<li><code><a title="smeagol.enrich.enrich_in_sliding_windows" href="#smeagol.enrich.enrich_in_sliding_windows">enrich_in_sliding_windows</a></code></li>
<li><code><a title="smeagol.enrich.enrich_in_window" href="#smeagol.enrich.enrich_in_window">enrich_in_window</a></code></li>
<li><code><a title="smeagol.enrich.enrich_over_shuffled" href="#smeagol.enrich.enrich_over_shuffled">enrich_over_shuffled</a></code></li>
<li><code><a title="smeagol.enrich.examine_thresholds" href="#smeagol.enrich.examine_thresholds">examine_thresholds</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>