<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>smeagol API documentation</title>
<meta name="description" content="SMEAGOL is a user-friendly Python library to identify and visualize enrichment (or depletion) of sequence motifs in nucleic acid sequences â€¦" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<link rel="preconnect" href="https://www.google.com">
<script async src="https://cse.google.com/cse.js?cx=017837193012385208679:pey8ky8gdqw"></script>
<style>
.gsc-control-cse {padding:0 !important;margin-top:1em}
body.gsc-overflow-hidden #sidebar {overflow: visible;}
</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Package <code>smeagol</code></h1>
</header>
<section id="section-intro">
<p>SMEAGOL is a user-friendly Python library to identify and visualize enrichment (or depletion) of sequence motifs in nucleic acid sequences. </p>
<p>SMEAGOL can load nucleic acid sequences as well as sequence motifs that represent the binding preferences of nucleic-acid binding proteins. These motifs can be analyzed and processed, and converted into Position Weight Matrices (PWMs). SMEAGOL further allows scanning of nucleic acid sequences with these motifs and scores each position of the sequence according to its match to the supplied motifs. It offers a statistical test for whether a given motif is significantly enriched or depleted in a sequence relative to a background model. Finally, SMEAGOL can load sequence variants and calculate their impact on the motif match score in their surrounding sequence.</p>
<p>The following sub-modules are included in this library.</p>
<ul>
<li>smeagol.aggregate: Aggregate results across different genomes.</li>
<li>smeagol.encode: Encoding nucleic acid sequences into numeric form for PWM scanning.</li>
<li>smeagol.enrich: Calculate enrichment or depletion of PWM matches in a sequence.</li>
<li>smeagol.io: Read and write data (motifs and sequences).</li>
<li>smeagol.matrices: Analyze motifs in the form of PFMs, PPMs and PWMs.</li>
<li>smeagol.models: Encoding of PWMs into a model to scan sequences.</li>
<li>smeagol.scan: Scan nucleic acid sequences with PWMs and score putative binding sites.</li>
<li>smeagol.utils: Miscellaneous functions used by other modules.</li>
<li>smeagol.variant: Predict the effects of sequence variants on the PWM match score.</li>
<li>smeagol.visualize: Generating visualizations.</li>
</ul>
<p>The individual functions in each sub-module are documented below.</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;
SMEAGOL is a user-friendly Python library to identify and visualize enrichment (or depletion) of sequence motifs in nucleic acid sequences. 

SMEAGOL can load nucleic acid sequences as well as sequence motifs that represent the binding preferences of nucleic-acid binding proteins. These motifs can be analyzed and processed, and converted into Position Weight Matrices (PWMs). SMEAGOL further allows scanning of nucleic acid sequences with these motifs and scores each position of the sequence according to its match to the supplied motifs. It offers a statistical test for whether a given motif is significantly enriched or depleted in a sequence relative to a background model. Finally, SMEAGOL can load sequence variants and calculate their impact on the motif match score in their surrounding sequence.

The following sub-modules are included in this library.

- smeagol.aggregate: Aggregate results across different genomes.
- smeagol.encode: Encoding nucleic acid sequences into numeric form for PWM scanning.
- smeagol.enrich: Calculate enrichment or depletion of PWM matches in a sequence.
- smeagol.io: Read and write data (motifs and sequences).
- smeagol.matrices: Analyze motifs in the form of PFMs, PPMs and PWMs.
- smeagol.models: Encoding of PWMs into a model to scan sequences.
- smeagol.scan: Scan nucleic acid sequences with PWMs and score putative binding sites.
- smeagol.utils: Miscellaneous functions used by other modules.
- smeagol.variant: Predict the effects of sequence variants on the PWM match score.
- smeagol.visualize: Generating visualizations.

The individual functions in each sub-module are documented below.

&#34;&#34;&#34;</code></pre>
</details>
</section>
<section>
<h2 class="section-title" id="header-submodules">Sub-modules</h2>
<dl>
<dt><code class="name"><a title="smeagol.aggregate" href="aggregate.html">smeagol.aggregate</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.encode" href="encode.html">smeagol.encode</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.enrich" href="enrich.html">smeagol.enrich</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.io" href="io.html">smeagol.io</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.matrices" href="matrices.html">smeagol.matrices</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.models" href="models.html">smeagol.models</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.scan" href="scan.html">smeagol.scan</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.utils" href="utils.html">smeagol.utils</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.variant" href="variant.html">smeagol.variant</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="smeagol.visualize" href="visualize.html">smeagol.visualize</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<div class="gcse-search" style="height: 70px"
data-as_oq="inurl:github.com/gruber-sciencelab/SMEAGOL site:gruber-sciencelab.github.io/SMEAGOL/"
data-gaCategoryParameter="smeagol">
</div>
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3><a href="#header-submodules">Sub-modules</a></h3>
<ul>
<li><code><a title="smeagol.aggregate" href="aggregate.html">smeagol.aggregate</a></code></li>
<li><code><a title="smeagol.encode" href="encode.html">smeagol.encode</a></code></li>
<li><code><a title="smeagol.enrich" href="enrich.html">smeagol.enrich</a></code></li>
<li><code><a title="smeagol.io" href="io.html">smeagol.io</a></code></li>
<li><code><a title="smeagol.matrices" href="matrices.html">smeagol.matrices</a></code></li>
<li><code><a title="smeagol.models" href="models.html">smeagol.models</a></code></li>
<li><code><a title="smeagol.scan" href="scan.html">smeagol.scan</a></code></li>
<li><code><a title="smeagol.utils" href="utils.html">smeagol.utils</a></code></li>
<li><code><a title="smeagol.variant" href="variant.html">smeagol.variant</a></code></li>
<li><code><a title="smeagol.visualize" href="visualize.html">smeagol.visualize</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>