
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Introduction &#8212; SMEAGOL 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="smeagol" href="modules.html" />
    <link rel="prev" title="Welcome to SMEAGOL’s documentation!" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>SMEAGOL is a user-friendly Python library to identify and visualize enrichment (or depletion) of sequence motifs in nucleic acid sequences.</p>
<p>SMEAGOL can load nucleic acid sequences as well as sequence motifs that represent the binding preferences of nucleic-acid binding proteins. These motifs can be analyzed and processed, and converted into Position Weight Matrices (PWMs). SMEAGOL further allows scanning of nucleic acid sequences with these motifs and scores each position of the sequence according to its match to the supplied motifs. It offers a statistical test for whether a given motif is significantly enriched or depleted in a sequence relative to a background model. Finally, SMEAGOL can load sequence variants and calculate their impact on the motif match score in their surrounding sequence.</p>
<p>The following sub-modules are included in this library.</p>
<ul class="simple">
<li><p>smeagol.aggregate: Aggregate results across different genomes.</p></li>
<li><p>smeagol.encode: Encoding nucleic acid sequences into numeric form for PWM scanning.</p></li>
<li><p>smeagol.enrich: Calculate enrichment or depletion of PWM matches in a sequence.</p></li>
<li><p>smeagol.io: Read and write data (motifs and sequences).</p></li>
<li><p>smeagol.matrices: Analyze motifs in the form of PFMs, PPMs and PWMs.</p></li>
<li><p>smeagol.models: Encoding of PWMs into a model to scan sequences.</p></li>
<li><p>smeagol.scan: Scan nucleic acid sequences with PWMs and score putative binding sites.</p></li>
<li><p>smeagol.utils: Miscellaneous functions used by other modules.</p></li>
<li><p>smeagol.variant: Predict the effects of sequence variants on the PWM match score.</p></li>
<li><p>smeagol.visualize: Generating visualizations.</p></li>
</ul>
<p>The individual functions in each sub-module are documented below.</p>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">SMEAGOL</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">smeagol</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Welcome to SMEAGOL’s documentation!</a></li>
      <li>Next: <a href="modules.html" title="next chapter">smeagol</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2022, Gruber Lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.5.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/readme.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>