<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>smeagol.utils API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<link rel="preconnect" href="https://www.google.com">
<script async src="https://cse.google.com/cse.js?cx=017837193012385208679:pey8ky8gdqw"></script>
<style>
.gsc-control-cse {padding:0 !important;margin-top:1em}
body.gsc-overflow-hidden #sidebar {overflow: visible;}
</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>smeagol.utils</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">import numpy as np
from ushuffle import shuffle, set_seed

# Biopython imports
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord

# Smeagol imports
from smeagol.io import read_fasta, write_fasta


def _equals(x, y, eps=1e-4):
    &#34;&#34;&#34;Function to test whether two values or lists/arrays are equal with some tolerance.
    
    Args:
        x, y: float, int, list or np.array objects
        eps: tolerance value
        
    Returns:
        True if x and y are equal within the tolerance limit, False otherwise
        
    &#34;&#34;&#34;
    if type(x) == type(y) == np.ndarray:
        assert x.shape == y.shape
        return np.all(abs(x - y) &lt; (eps * x.size))
    elif type(x) == type(y) == list:
        assert len(x) == len(y)
        return np.all(abs(x - y) &lt; (eps * len(x)))
    else:
        return abs(x - y) &lt; eps


def shuffle_records(records, simN, simK, out_file=None, seed=1):
    &#34;&#34;&#34;Function to shuffle sequences.
    
    Args:
        records (list): list of seqRecord objects
        simN (int): Number of times to shuffle
        simK (int): k-mer frequency to conserve
        out_file (str): Path to output file (optional)
    
    Returns:
        shuf_records (list): list of shuffled sequences
        Writes shuf_records to file if out_file provided.

    &#34;&#34;&#34;
    shuf_records = []
    
    if seed is not None:
        set_seed(seed)
    
    # Shuffle each record
    for record in records:
        for n in range(simN):
            new_seq = shuffle(str.encode(record.seq.__str__()), simK).decode()
            new_seq = SeqRecord(Seq(new_seq), id=&#34;background_seq_{0:d}&#34;.format(n))
            new_seq.name = record.id
            shuf_records.append(new_seq)                
    print(&#39;Shuffled {} sequence(s) {} times while conserving k-mer frequency for k = {}.&#39;.format(len(records), simN, simK))

    # Write to file
    if out_file is not None:
        write_fasta(shuf_records, out_file)

    return shuf_records


def get_Seq(seq):
    &#34;&#34;&#34;Function to convert a sequence into a Seq object.
    
    Args:
        seq (str, Seq or SeqRecord): sequence
        
    Returns:
        Seq object
        
    &#34;&#34;&#34;
    if type(seq) == str:
        return Seq(seq)
    elif type(seq) == SeqRecord:
        return seq.seq
    elif type(seq) == Seq:
        return seq
    else:
        raise TypeError(&#39;Input sequence must be a string, Seqrecord or Seq object.&#39;)
        
        
def read_bg_seqs(file, records, simN):
    &#34;&#34;&#34;Function to read background sequences from FASTA file.
    
    Args:
        file (str): Path to fasta (or fasta.gz) file
        records (list): Original un-shuffled records
        simN (int): Number of shuffles
        
    Returns:
        list of seqrecord objects
    
    &#34;&#34;&#34;
    # Read background sequences
    bg = read_fasta(file)
    # Check number of sequences
    assert len(bg) == len(records) * simN
    # Check sequence length and fix name
    for i, bg_record in enumerate(bg):
        matching_record = records[i // simN]
        assert len (bg_record) == len(matching_record), &#34;length of shuffled sequence does not match original sequence.&#34;
        bg_record.name = matching_record.name
    return bg
    
    
def _get_tiling_windows_over_record(record, width, shift):
    &#34;&#34;&#34;Function to get tiling windows over a sequence.
    
    Args:
        record (SeqRecord): SeqRecord object
        width (int): width of tiling windows
        shift (int): shift between successive windows
    
    Returns:
        windows (pd.DataFrame): windows covering input sequence
        
    &#34;&#34;&#34;
    # Get start positions
    starts = list(range(0, len(record), shift))
    # Get end positions
    ends = [np.min([x + width, len(record)]) for x in starts]
    # Add sequence ID
    idxs = np.tile(record.id, len(starts))
    # Combine
    windows = pd.DataFrame({&#39;id&#39;:idxs, &#39;start&#39;:starts, &#39;end&#39;:ends})
    return windows


def get_tiling_windows_over_genome(genome, width, shift=None):
    &#34;&#34;&#34;Function to get tiling windows over a genome.
        
    Args:
        genome (list): list of SeqRecord objects
        width (int): width of tiling windows
        shift (int): shift between successive windows. By default the same as width.
    
    Returns:
        windows (pd.DataFrame): windows covering all sequences in genome
        
    &#34;&#34;&#34;
    if shift is None:
        shift = width
    if len(genome) == 1:
        windows = _get_tiling_windows_over_record(genome[0], width, shift)
    else:
        windows = pd.concat([_get_tiling_windows_over_record(record, width, shift) for record in genome])
    return windows


def get_site_seq(site, seqs):
    &#34;&#34;&#34;Function to extract the sequence for a binding site.
    
    Args:
        site: dataframe containing name, start and end.
        seqs (list / str): list of seqrecord objects or fasta file.
    Returns:
        result (str): sequence of binding site
    &#34;&#34;&#34;
    if type(seqs) == str:
        seqs = read_fasta(seqs)
    seq = [x for x in seqs if x.name == site[&#39;name&#39;]][0]
    return str(seq.seq[site.start:site.end])</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="smeagol.utils.get_Seq"><code class="name flex">
<span>def <span class="ident">get_Seq</span></span>(<span>seq)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to convert a sequence into a Seq object.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>seq</code></strong> :&ensp;<code>str, Seq</code> or <code>SeqRecord</code></dt>
<dd>sequence</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>Seq object</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_Seq(seq):
    &#34;&#34;&#34;Function to convert a sequence into a Seq object.
    
    Args:
        seq (str, Seq or SeqRecord): sequence
        
    Returns:
        Seq object
        
    &#34;&#34;&#34;
    if type(seq) == str:
        return Seq(seq)
    elif type(seq) == SeqRecord:
        return seq.seq
    elif type(seq) == Seq:
        return seq
    else:
        raise TypeError(&#39;Input sequence must be a string, Seqrecord or Seq object.&#39;)</code></pre>
</details>
</dd>
<dt id="smeagol.utils.get_site_seq"><code class="name flex">
<span>def <span class="ident">get_site_seq</span></span>(<span>site, seqs)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to extract the sequence for a binding site.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>site</code></strong></dt>
<dd>dataframe containing name, start and end.</dd>
</dl>
<p>seqs (list / str): list of seqrecord objects or fasta file.</p>
<h2 id="returns">Returns</h2>
<p>result (str): sequence of binding site</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_site_seq(site, seqs):
    &#34;&#34;&#34;Function to extract the sequence for a binding site.
    
    Args:
        site: dataframe containing name, start and end.
        seqs (list / str): list of seqrecord objects or fasta file.
    Returns:
        result (str): sequence of binding site
    &#34;&#34;&#34;
    if type(seqs) == str:
        seqs = read_fasta(seqs)
    seq = [x for x in seqs if x.name == site[&#39;name&#39;]][0]
    return str(seq.seq[site.start:site.end])</code></pre>
</details>
</dd>
<dt id="smeagol.utils.get_tiling_windows_over_genome"><code class="name flex">
<span>def <span class="ident">get_tiling_windows_over_genome</span></span>(<span>genome, width, shift=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to get tiling windows over a genome.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>genome</code></strong> :&ensp;<code>list</code></dt>
<dd>list of SeqRecord objects</dd>
<dt><strong><code>width</code></strong> :&ensp;<code>int</code></dt>
<dd>width of tiling windows</dd>
<dt><strong><code>shift</code></strong> :&ensp;<code>int</code></dt>
<dd>shift between successive windows. By default the same as width.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>windows (pd.DataFrame): windows covering all sequences in genome</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_tiling_windows_over_genome(genome, width, shift=None):
    &#34;&#34;&#34;Function to get tiling windows over a genome.
        
    Args:
        genome (list): list of SeqRecord objects
        width (int): width of tiling windows
        shift (int): shift between successive windows. By default the same as width.
    
    Returns:
        windows (pd.DataFrame): windows covering all sequences in genome
        
    &#34;&#34;&#34;
    if shift is None:
        shift = width
    if len(genome) == 1:
        windows = _get_tiling_windows_over_record(genome[0], width, shift)
    else:
        windows = pd.concat([_get_tiling_windows_over_record(record, width, shift) for record in genome])
    return windows</code></pre>
</details>
</dd>
<dt id="smeagol.utils.read_bg_seqs"><code class="name flex">
<span>def <span class="ident">read_bg_seqs</span></span>(<span>file, records, simN)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to read background sequences from FASTA file.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>file</code></strong> :&ensp;<code>str</code></dt>
<dd>Path to fasta (or fasta.gz) file</dd>
<dt><strong><code>records</code></strong> :&ensp;<code>list</code></dt>
<dd>Original un-shuffled records</dd>
<dt><strong><code>simN</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of shuffles</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>list of seqrecord objects</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def read_bg_seqs(file, records, simN):
    &#34;&#34;&#34;Function to read background sequences from FASTA file.
    
    Args:
        file (str): Path to fasta (or fasta.gz) file
        records (list): Original un-shuffled records
        simN (int): Number of shuffles
        
    Returns:
        list of seqrecord objects
    
    &#34;&#34;&#34;
    # Read background sequences
    bg = read_fasta(file)
    # Check number of sequences
    assert len(bg) == len(records) * simN
    # Check sequence length and fix name
    for i, bg_record in enumerate(bg):
        matching_record = records[i // simN]
        assert len (bg_record) == len(matching_record), &#34;length of shuffled sequence does not match original sequence.&#34;
        bg_record.name = matching_record.name
    return bg</code></pre>
</details>
</dd>
<dt id="smeagol.utils.shuffle_records"><code class="name flex">
<span>def <span class="ident">shuffle_records</span></span>(<span>records, simN, simK, out_file=None, seed=1)</span>
</code></dt>
<dd>
<div class="desc"><p>Function to shuffle sequences.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>records</code></strong> :&ensp;<code>list</code></dt>
<dd>list of seqRecord objects</dd>
<dt><strong><code>simN</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of times to shuffle</dd>
<dt><strong><code>simK</code></strong> :&ensp;<code>int</code></dt>
<dd>k-mer frequency to conserve</dd>
<dt><strong><code>out_file</code></strong> :&ensp;<code>str</code></dt>
<dd>Path to output file (optional)</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>shuf_records (list): list of shuffled sequences
Writes shuf_records to file if out_file provided.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def shuffle_records(records, simN, simK, out_file=None, seed=1):
    &#34;&#34;&#34;Function to shuffle sequences.
    
    Args:
        records (list): list of seqRecord objects
        simN (int): Number of times to shuffle
        simK (int): k-mer frequency to conserve
        out_file (str): Path to output file (optional)
    
    Returns:
        shuf_records (list): list of shuffled sequences
        Writes shuf_records to file if out_file provided.

    &#34;&#34;&#34;
    shuf_records = []
    
    if seed is not None:
        set_seed(seed)
    
    # Shuffle each record
    for record in records:
        for n in range(simN):
            new_seq = shuffle(str.encode(record.seq.__str__()), simK).decode()
            new_seq = SeqRecord(Seq(new_seq), id=&#34;background_seq_{0:d}&#34;.format(n))
            new_seq.name = record.id
            shuf_records.append(new_seq)                
    print(&#39;Shuffled {} sequence(s) {} times while conserving k-mer frequency for k = {}.&#39;.format(len(records), simN, simK))

    # Write to file
    if out_file is not None:
        write_fasta(shuf_records, out_file)

    return shuf_records</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<div class="gcse-search" style="height: 70px"
data-as_oq="inurl:github.com/gruber-sciencelab/SMEAGOL site:gruber-sciencelab.github.io/SMEAGOL/"
data-gaCategoryParameter="smeagol.utils">
</div>
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="smeagol" href="index.html">smeagol</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="smeagol.utils.get_Seq" href="#smeagol.utils.get_Seq">get_Seq</a></code></li>
<li><code><a title="smeagol.utils.get_site_seq" href="#smeagol.utils.get_site_seq">get_site_seq</a></code></li>
<li><code><a title="smeagol.utils.get_tiling_windows_over_genome" href="#smeagol.utils.get_tiling_windows_over_genome">get_tiling_windows_over_genome</a></code></li>
<li><code><a title="smeagol.utils.read_bg_seqs" href="#smeagol.utils.read_bg_seqs">read_bg_seqs</a></code></li>
<li><code><a title="smeagol.utils.shuffle_records" href="#smeagol.utils.shuffle_records">shuffle_records</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>